pkg/scalers/azure_log_analytics_scaler.go

```diff
@@ -0,0 +1,490 @@
+	receivedMetric, err := s.getMetricData()
+
+	if err != nil {
+		logAnalyticsLog.Error(err, "Failed to get metrics.")
+		return []external_metrics.ExternalMetricValue{}, err
+	}
+
+	metric := external_metrics.ExternalMetricValue{
+		MetricName: metricName,
+		Value:      *resource.NewQuantity(receivedMetric.value, resource.DecimalSI),
+		Timestamp:  metav1.Now(),
+	}
+
+	return append([]external_metrics.ExternalMetricValue{}, metric), nil
+}
+
+func (s *azureLogAnalyticsScaler) Close() error {
+	return nil
+}
+
+func (s *azureLogAnalyticsScaler) updateCache() error {
+	if s.cache.metricValue < 0 {
+		receivedMetric, err := s.getMetricData()
+
+		if err != nil {
+			logAnalyticsLog.Error(err, "Error processing Log Analytics query")
+			return err
+		}
+
+		s.cache.metricValue = receivedMetric.value
+
+		if receivedMetric.threshold > 0 {
+			s.cache.metricThreshold = receivedMetric.threshold
+		} else {
+			s.cache.metricThreshold = s.metadata.threshold
+		}
+	}
+
+	return nil
+}
+
+func (s *azureLogAnalyticsScaler) getMetricData() (metricsData, error) {
+	tokenInfo, err := s.getAccessToken()
+	if err != nil {
+		return metricsData{}, err
+	}
+
+	metricsInfo, err := s.executeQuery(s.metadata.query, tokenInfo)
+	if err != nil {
+		return metricsData{}, err
+	}
+
+	logAnalyticsLog.V(1).Info("Getting metrics value", "metrics value", metricsInfo.value)
+	return metricsInfo, nil
+}
+
+func (s *azureLogAnalyticsScaler) getAccessToken() (tokenData, error) {
+	//if there is no token yet or it will be expired in less, that 30 secs
+	currentTimeSec := time.Now().Unix()
+	if currentTimeSec+30 > s.cache.accessToken.ExpiresOn {
+		tokenInfo, err := s.refreshAccessToken()
+		if err != nil {
+			return tokenData{}, err
+		}
+
+		s.cache.accessToken = tokenInfo
+		return tokenInfo, nil
+	}
+	return s.cache.accessToken, nil
+}
+
+func (s *azureLogAnalyticsScaler) executeQuery(query string, tokenInfo tokenData) (metricsData, error) {
+	queryData := queryResult{}
+
+	body, statusCode, err := s.executeLogAnalyticsREST(query, tokenInfo)
+
+	if body == nil || len(body) == 0 {
+		return metricsData{}, fmt.Errorf("Error executing Log Analytics REST API request: empty body")
+	}
+
+	//Handle expired token
+	if statusCode == 403 || strings.Contains(string(body), "TokenExpired") {
+		logAnalyticsLog.Info("Token expired, refreshing token...")
+
+		tokenInfo, err := s.refreshAccessToken()
+
+		if err == nil {
+			body, statusCode, err = s.executeLogAnalyticsREST(query, tokenInfo)
+		} else {
+			return metricsData{}, err
+		}
+	}
+
+	if err != nil {
+		return metricsData{}, err
+	}
+
+	err = json.NewDecoder(bytes.NewReader(body)).Decode(&queryData)
+	if err != nil {
+		if statusCode != 200 {
+			return metricsData{}, fmt.Errorf("Error processing Log Analytics REST API request: %d", statusCode)
+		}
+		return metricsData{}, fmt.Errorf("Can't decode JSON from Log Analytics REST API result: %v", err)
+	}
+
+	if statusCode == 200 {
```

Comment:
nit: can this be changed into an early return. i.e: 

```
if statusCode != 200 {
    return .....
}
```

Reply:
Thanks for catching. Moved it upper in last commit.